### ISSUE TITLE
bug report: "TROUBLE in pre_loop_hook: BDB can't create db env. at /var/lib/amavis/db: Permission denied"

### ISSUE THREAD

**Issue by Eirikr70** [docker-mailserver/docker-mailserver#4202]
### 📝 Preliminary Checks

- [X] I tried searching for an existing issue and followed the [debugging docs](https://docker-mailserver.github.io/docker-mailserver/latest/config/debugging/) advice, but still need assistance.

### 👀 What Happened?

When activating Amavis, I always get the following error, and Amavis dies and restarts
```
mailserver  | 2024-09-28T17:48:47.868966+02:00 mail amavis[956]: (!!)TROUBLE in pre_loop_hook: BDB can't create db env. at /var/lib/amavis/db: Permission denied, Permission denied. at /usr/share/perl5/Amavis/DB.pm line 58.
mailserver  | 2024-09-28 17:48:47,881 WARN exited: amavis (exit status 255; not expected)
```

### 👟 Reproduction Steps

Set `ENABLE_AMAVIS=1` in `mailserver.env`

### 🐋 DMS Version

v14.0.0

### 💻 Operating System and Architecture

Debian 12

### ⚙️ Container configuration files

```yml
services:
  mailserver:
    image: docker.io/mailserver/docker-mailserver:latest
    container_name: mailserver
    # If the FQDN for your mail-server is only two labels (eg: example.com),
    # you can assign this entirely to `hostname` and remove `domainname`.
    hostname: mail.xxx.fr
    domainname: xxx.fr
    env_file: mailserver.env
    # More information about the mail-server ports:
    # https://docker-mailserver.github.io/docker-mailserver/edge/config/security/understanding-the-ports/
    # To avoid conflicts with yaml base-60 float, DO NOT remove the quotation marks.
    ports:
      - "25:25"    # SMTP  (explicit TLS => STARTTLS)
#      - "143:143"  # IMAP4 (explicit TLS => STARTTLS)
      - "465:465"  # ESMTP (implicit TLS)
#      - "587:587"  # ESMTP (explicit TLS => STARTTLS)
      - "993:993"  # IMAP4 (implicit TLS)
    volumes:
      - /srv/nas/mail/dms/mail-data/:/var/mail/
      - /srv/nas/mail/dms/mail-state/:/var/mail-state/
      - /srv/nas/logs:/var/log/mail
      - /srv/nas/mail/dms/config/:/tmp/docker-mailserver/
      - /etc/localtime:/etc/localtime:ro
      - /srv/nas/mail/fail2ban/:/etc/fail2ban/
#      - /srv/nas/mail/amavis:/var/lib/amavis
      - /home/eric/swag/config/etc/letsencrypt:/etc/letsencrypt:ro
    restart: always
    stop_grace_period: 1m
    cap_add:
      - NET_ADMIN
      - SYS_PTRACE
#    runtime: "runsc"

# volumes:
#   logs:
#     external: true
```

### 📜 Relevant log output

```Text
mailserver  | 2024-09-28 17:48:40+02:00 INFO  start-mailserver.sh: Welcome to docker-mailserver v14.0.0
mailserver  | 2024-09-28 17:48:40+02:00 INFO  start-mailserver.sh: Checking configuration
mailserver  | 2024-09-28 17:48:40+02:00 INFO  start-mailserver.sh: Configuring mail server
mailserver  | 2024-09-28 17:48:43+02:00 INFO  start-mailserver.sh: Starting daemons
mailserver  | 2024-09-28 17:48:45+02:00 INFO  start-mailserver.sh: mail.xxx.fr is up and running
mailserver  | 2024-09-28T17:48:47.227092+02:00 mail amavis[956]: starting. /usr/sbin/amavisd at mail.xxx.fr amavis-2.13.0 (20230106), Unicode aware, LC_CTYPE="C.UTF-8"
mailserver  | 2024-09-28T17:48:47.227119+02:00 mail amavis[956]: perl=5.036000, user=, EUID: 999 (999);  group=(), EGID: 999 999 (999 999)
mailserver  | 2024-09-28T17:48:47.815512+02:00 mail amavis[956]: Net::Server: Group Not Defined.  Defaulting to EGID '999 999'
mailserver  | 2024-09-28T17:48:47.815578+02:00 mail amavis[956]: Net::Server: User Not Defined.  Defaulting to EUID '999'
mailserver  | 2024-09-28T17:48:47.826402+02:00 mail amavis[956]: No ext program for   .zoo, tried: zoo
mailserver  | 2024-09-28T17:48:47.826455+02:00 mail amavis[956]: No ext program for   .doc, tried: ripole
mailserver  | 2024-09-28T17:48:47.826855+02:00 mail amavis[956]: No ext program for   .zst, tried: unzstd
mailserver  | 2024-09-28T17:48:47.826901+02:00 mail amavis[956]: No decoder for       .F   
mailserver  | 2024-09-28T17:48:47.826919+02:00 mail amavis[956]: No decoder for       .doc 
mailserver  | 2024-09-28T17:48:47.826936+02:00 mail amavis[956]: No decoder for       .zoo 
mailserver  | 2024-09-28T17:48:47.826951+02:00 mail amavis[956]: No decoder for       .zst 
mailserver  | 2024-09-28T17:48:47.868966+02:00 mail amavis[956]: (!!)TROUBLE in pre_loop_hook: BDB can't create db env. at /var/lib/amavis/db: Permission denied, Permission denied. at /usr/share/perl5/Amavis/DB.pm line 58.
mailserver  | 2024-09-28 17:48:47,881 WARN exited: amavis (exit status 255; not expected)
```

**Comment by polarathene** [comment]
> `/srv/nas`

This is a NFS / SMB mount point correct?

Please share more context about your NAS environment. Is it also Debian 12? Kernel? NFS version?

I don't know the specific cause, and I'm not likely able to assist you here as remote file systems and NAS products are known to be problematic with DMS. The file path that is failing is relying on that remote mount point, thus is where the incompatibility is coming from.

Mount this locally instead, or use a named volume (this will also be local):

```yaml
- /srv/nas/mail/dms/mail-state/:/var/mail-state/
```

That will solve the issue. You can also try `ENABLE_AMAVIS=0` + `ENABLE_RSPAMD=1` to switch to rspamd (future successor service to replace amavis and others), and logs will mention several other services you can disable via ENV related to this switch to have rspamd handle it all. That may work for you without shifting DMS service state to your Docker host.

Presumably given the type of file it tried to use, this might be related to file locks (_in the past a contributor moved our internal scripts locking from `flock` to file-based locks to work around this issue, but we cannot do this for the services that DMS runs_), or another common one is ownership / permissions being problematic (_such as the container root user not being allowed to change such on the remote filesystem end_), there can also be issues with cache or kernel compatibility along with configuration IIRC.

It's really unclear as typically our users with these problems have the remote filesystem for convenience without much sysadmin knowledge about proper troubleshooting or configuration in these scenarios (_as it works good enough for most things elsewhere_).

**Comment by Eirikr70** [comment]
Hi, thanks for your fast answer. `/srv/nas` is just the name of a hard drive. My system is set on one drive and all my data (Docker volumes) on another, which I named NAS. It is all on the same host on Debian 12.
Trying with RSPAMD enabled. It seems to run fine. Thanks for the advice. I'll have to learn about Rspamd, which I hadn't heard about.
Got to understand the advice the logs give about OpenDKIM and OpenDMARC ...

**Comment by polarathene** [comment]
> Trying with RSPAMD enabled. It seems to run fine. Thanks for the advice. I'll have to learn about Rspamd, which I hadn't heard about.

Yeah, I'm not too familiar with it myself, but another maintainer takes care of it and is rather confident that it will replace many of our existing services and do a better job. Other projects similar to DMS already use rspamd for this.

You can look over our current rspamd docs and provide any feedback or questions related to that if you like so that we can know where to improve it for future users benefit :)

> Got to understand the advice the logs give about OpenDKIM and OpenDMARC ...

That's likely about disabling these in favor of rspamd. I have a small snippet recently that shows the related ENV changes:

https://github.com/docker-mailserver/docker-mailserver/issues/4189#issuecomment-2378245329

So you probably want these:

```env
ENABLE_RSPAMD=1
# Enabling rspamd replaces these services, disable them:
ENABLE_AMAVIS=0
ENABLE_OPENDKIM=0
ENABLE_OPENDMARC=0
ENABLE_POLICYD_SPF=0
```

@georglauterbach or our docs probably describe this better, but afaik that's at least the minimum requirement to switch over to rspamd.

You may need to manage some configuration for rspamd, I'm aware that our DKIM support differs, so any DKIM keys would need to be relocated or regenerated, and potentially provide your own rspamd DKIM config. If you don't have DKIM setup yet, then that shouldn't be anything to worry about AFAIK 👍

**Comment by Eirikr70** [comment]
Thanks for telling me about re-generating the DKIM keys. Got 10/10 on mail-tester.com and saving my deliverability thanks to you !

**Comment by Eirikr70** [comment]
Just so you know : the link for Abusix account creation seems dead. I'll try to get the new one for you if a- there is one, b- I find it.

**Comment by georglauterbach** [comment]
@polarathene is right about the ENV, and our docs cover these as well. From what you wrote @Eirikr70, I assume you also got DKIM working. This issue appears to be solved to me :)

**Comment by Eirikr70** [comment]
> @polarathene is right about the ENV, and our docs cover these as well. From what you wrote @Eirikr70, I assume you also got DKIM working. This issue appears to be solved to me :)

It is. Thank you !

**Comment by polarathene** [comment]
> This issue appears to be solved to me :)

Do we need a new issue for tracking blockers for rspamd switching to default? What comes to mind is:
- [ ] Migration docs (private keys + config)
- [ ] ValKey replacement of Redis (_only affects users relying on internal service, assuming some migration would be required and external Redis configurations would be unaffected?_), should be done before the switch to minimize impact (assuming there is any)
- [ ] Switch to `selector_map` for rspamd (_simplifies DKIM scripts_).
- [ ] Unified DKIM generation (optional/related). This could generate OpenDKIM/Rspamd config at runtime and manage keys at the same location + generation script, it would make the docs migration effectively redundant (_existing users will likely need to move keys to new location or regenerate and update DNS_).

**Comment by georglauterbach** [comment]
I tagged this issue; I think we do not need a new tracking issue.

**Comment by Eirikr70** [comment]
> I tagged this issue; I think we do not need a new tracking issue.

As the author of this issue, I couldn't agree more ! ;)
